OSSM-3987 Increase test coverage cluster wide

fjglira opened the pull request:
Jira: https://issues.redhat.com/browse/OSSM-3987

Adding new test case to increase test coverage:

- Check Rolebindings: https://issues.redhat.com/browse/OSSM-3468

fjglira commented:
Work done and ready to review

fjglira commented:
Adding change in sub test case: cluster wide works with profiles. Going to use this profile: `gateway-controller` to meet this Jira: https://issues.redhat.com/browse/OSSM-3562

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
+			t.LogStep("Check that Rolebindings are not created in the member namespaces")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				oc.Get(t, "member-0", "rolebindings", "",
+					assert.OutputContains("system:deployers",
+						"The Rolebings contains system:deployers role",
+						"The Rolebings does not contains system:deployers role"),
+					assert.OutputContains("system:image-builders",
+						"The Rolebings contains system:image-builders role",
+						"The Rolebings does not contains system:image-builders role"),
+					assert.OutputContains("system:image-pullers",
+						"The Rolebings contains system:image-pullers role",
+						"The Rolebings does not contains system:image-pullers role"),
```
Why check these rolebindings? They're not related to OSSM, so I think we should ignore them.

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
+						"The Rolebings contains system:image-pullers role",
+						"The Rolebings does not contains system:image-pullers role"),
+					assert.OutputDoesNotContain("istiod-clusterrole-basic-istio-system",
+						"The Rolebings does not contains istiod-clusterrole-basic-istio-system role",
+						"The Rolebings contains istiod-clusterrole-basic-istio-system role"),
```
```suggestion
						"The istiod-clusterrole-basic-istio-system RoleBinding does not exist",
						"The istiod-clusterrole-basic-istio-system RoleBinding exists"),
```

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
+						"The Rolebings does not contains istiod-clusterrole-basic-istio-system role",
+						"The Rolebings contains istiod-clusterrole-basic-istio-system role"),
+					assert.OutputDoesNotContain("istiod-gateway-controller-basic-istio-system",
+						"The Rolebings does not contains istiod-gateway-controller-basic-istio-system role",
+						"The Rolebings contains istiod-gateway-controller-basic-istio-system role"))
```
Same here

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -70,6 +70,29 @@ func TestClusterWideMode(t *testing.T) {
+			})
+		})
+
 		t.NewSubTest("customize SMMR").Run(func(t test.TestHelper) {
```
Shouldn't this be multiple subtests, since many different things are tested here? It's also hard to follow what this subtest does, because it's so long.

fjglira replied:
The only thing to not break in more sub test cases at the beginning is the dependency between the steps, but sure I can split in more subtest cases

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -104,6 +127,19 @@ func TestClusterWideMode(t *testing.T) {
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				shell.Execute(t,
+					fmt.Sprintf("oc -n %s get smmr default", meshNamespace),
+					require.OutputContains("1/1",
+						"one namespace are in members",
+						"expected SMMR to show 1 member namespace, but that wasn't the case"))
```
Why not check the actual namespace name instead of just the namespace count? The operator could add a completely different namespace here and the test would still pass.

luksa replied:
BTW, why do we create 50 member namespaces? Seems unnecessary. 5 would be more than enough.

fjglira replied:
Improved this

fjglira commented:
/test servicemesh-ocp4.13-lp-interop-images

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -172,33 +208,36 @@ func TestClusterWideMode(t *testing.T) {
+			t.Log("Check whether the cluster wide feature works with profiles")
+
+			t.LogStep("Delete SMCP and SMMR")
+			oc.RecreateNamespace(t, meshNamespace)
```
Note: deleting and recreating the namespace takes a long time because of Kiali. Whenever possible, delete the individual resources instead of the entire namespace.

fjglira replied:
Improved this

fjglira commented:
Imrpove execution time by non recreating the entire namespace

```
--- PASS: TestClusterWideMode (162.04s)
    --- PASS: TestClusterWideMode/SMMR_auto-creation (0.52s)
    --- PASS: TestClusterWideMode/default_namespace_selector (3.44s)
    --- PASS: TestClusterWideMode/RoleBindings_verification (0.52s)
    --- PASS: TestClusterWideMode/customize_SMMR (3.19s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_IN (3.25s)
    --- PASS: TestClusterWideMode/verify_multiple_memberselector (2.74s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_NOTIN (5.63s)
    --- PASS: TestClusterWideMode/verify_sidecar_injection (6.93s)
    --- PASS: TestClusterWideMode/cluster-scoped_watches_in_istiod (7.88s)
    --- PASS: TestClusterWideMode/cluster_wide_works_with_profiles (10.18s)
PASS
ok      github.com/maistra/maistra-test-tool/pkg/tests/ossm/operator    164.104s
```

fjglira commented:
@marko all the changes done

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -77,57 +101,70 @@ func TestClusterWideMode(t *testing.T) {
-					require.OutputContains("55/55",
-						"All the namespaces are in members except member-0",
-						"expected SMMR to show 55 member namespace, but that wasn't the case"))
+				verifyMembersInSMMR(t, meshNamespace, membersList, true)
+				verifyMembersInSMMR(t, meshNamespace, notInMembersList, false)
```
The function name is misleading when you pass in `false`. Please create two functions (e.g. `assertMembers`, `assertNonMembers` or something similar).

fjglira replied:
Done

luksa commented:
What's the purpose of the `clusterScoped` file? Why is it called that and not `clusterWide`? Also, why doesn't it have the `.yaml` extension?

fjglira commented:
> What's the purpose of the `clusterScoped` file? Why is it called that and not `clusterWide`? Also, why doesn't it have the `.yaml` extension?

Deleted, was used before but when I added to work with gateway profile was unused.

fjglira commented:
Tested
```
--- PASS: TestClusterWideMode (120.15s)
    --- PASS: TestClusterWideMode/SMMR_auto-creation (0.61s)
    --- PASS: TestClusterWideMode/default_namespace_selector (4.20s)
    --- PASS: TestClusterWideMode/RoleBindings_verification (0.63s)
    --- PASS: TestClusterWideMode/customize_SMMR (3.32s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_IN (3.38s)
    --- PASS: TestClusterWideMode/verify_multiple_memberselector (3.29s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_NOTIN (6.20s)
    --- PASS: TestClusterWideMode/verify_sidecar_injection (14.65s)
    --- PASS: TestClusterWideMode/cluster-scoped_watches_in_istiod (6.88s)
    --- PASS: TestClusterWideMode/cluster_wide_works_with_profiles (11.96s)
```

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -172,36 +198,80 @@ func TestClusterWideMode(t *testing.T) {
+		}
+		if found != shouldExist {
+			if shouldExist {
+				t.Fatalf("The member '%s' is missing from the members list.\n", member)
```
```suggestion
				t.Fatalf("The member '%s' is missing from the members list.", member)
```

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -172,36 +198,80 @@ func TestClusterWideMode(t *testing.T) {
+			if shouldExist {
+				t.Fatalf("The member '%s' is missing from the members list.\n", member)
+			} else {
+				t.Fatalf("The member '%s' exists in the members list.\n", member)
```
```suggestion
				t.Fatalf("The member '%s' exists in the members list.", member)
```

luksa commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -172,36 +198,80 @@ func TestClusterWideMode(t *testing.T) {
+
+// verifyMembersInSMMR verifies whether the SMMR has or not have the members provided in the members list
+func verifyMembersInSMMR(t test.TestHelper, meshNamespace string, membersList []string, shouldExist bool) {
+	smcpYaml := oc.GetYaml(t, meshNamespace, "smmr", "default")
```
Misleading name. It's a SMMR, not SMCP.

fjglira replied:
Done

jewertow commented on pkg/tests/ossm/operator/clusterwide_mode_test.go:
```diff
@@ -56,18 +64,33 @@ func TestClusterWideMode(t *testing.T) {
+			t.LogStep("Check that Rolebindings are not created in the member namespaces")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				oc.Get(t, "member-0", "rolebindings", "",
+					assert.OutputContains("prometheus-istio-system",
```
This role binding was removed, but then I [reverted](https://github.com/maistra/istio-operator/pull/1194) that change, so now this test should fail. Actually, since we will no longer use our prometheus, we can revert my commit to remove that role binding back, but it's too late and it has to be postponed to 2.4.1.
Anyway, this assertion should fail and you will have to remove it for now.

fjglira commented:
Test passed with latest changes
```
--- PASS: TestClusterWideMode (136.59s)
    --- PASS: TestClusterWideMode/SMMR_auto-creation (0.62s)
    --- PASS: TestClusterWideMode/default_namespace_selector (4.21s)
    --- PASS: TestClusterWideMode/RoleBindings_verification (0.63s)
    --- PASS: TestClusterWideMode/customize_SMMR (3.44s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_IN (3.43s)
    --- PASS: TestClusterWideMode/verify_multiple_memberselector (4.71s)
    --- PASS: TestClusterWideMode/verify_memberselector_operator_NOTIN (16.19s)
    --- PASS: TestClusterWideMode/verify_sidecar_injection (15.56s)
    --- PASS: TestClusterWideMode/cluster-scoped_watches_in_istiod (10.28s)
    --- PASS: TestClusterWideMode/cluster_wide_works_with_profiles (12.91s)
```
